packages/gatsby-plugin-gatsby-cloud/src/build-headers-program.js

```diff
@@ -0,0 +1,348 @@
+        (accum, header) => `${accum}  ${header}\n`,
+        ``
+      )
+      return `${text}${path}\n${headersString}`
+    },
+    ``
+  )
+}
+
+// program methods
+
+const validateUserOptions = (pluginOptions, reporter) => headers => {
+  if (!validHeaders(headers, reporter)) {
+    throw new Error(
+      `The "headers" option to gatsby-plugin-gatsby-cloud is in the wrong shape. ` +
+        `You should pass in a object with string keys (representing the paths) and an array ` +
+        `of strings as the value (representing the headers). ` +
+        `Check your gatsby-config.js.`
+    )
+  }
+
+  ;[`mergeSecurityHeaders`, `mergeLinkHeaders`, `mergeCachingHeaders`].forEach(
+    mergeOption => {
+      if (!_.isBoolean(pluginOptions[mergeOption])) {
+        throw new Error(
+          `The "${mergeOption}" option to gatsby-plugin-gatsby-cloud must be a boolean. ` +
+            `Check your gatsby-config.js.`
+        )
+      }
+    }
+  )
+
+  if (!_.isFunction(pluginOptions.transformHeaders)) {
+    throw new Error(
+      `The "transformHeaders" option to gatsby-plugin-gatsby-cloud must be a function ` +
+        `that returns an array of header strings. ` +
+        `Check your gatsby-config.js.`
+    )
+  }
+
+  return headers
+}
+
+const mapUserLinkHeaders = ({
+  manifest,
+  pathPrefix,
+  publicFolder,
+}) => headers =>
+  _.mapValues(headers, headerList =>
+    _.map(headerList, transformLink(manifest, publicFolder, pathPrefix))
+  )
+
+const mapUserLinkAllPageHeaders = (
+  pluginData,
+  { allPageHeaders }
+) => headers => {
+  if (!allPageHeaders) {
+    return headers
+  }
+
+  const { pages, manifest, publicFolder, pathPrefix } = pluginData
+
+  const headersList = _.map(
+    allPageHeaders,
+    transformLink(manifest, publicFolder, pathPrefix)
+  )
+
+  const duplicateHeadersByPage = {}
+  pages.forEach(page => {
+    const pathKey = headersPath(pathPrefix, page.path)
+    duplicateHeadersByPage[pathKey] = headersList
+  })
+
+  return defaultMerge(headers, duplicateHeadersByPage)
+}
+
+const applyLinkHeaders = (pluginData, { mergeLinkHeaders }) => headers => {
+  if (!mergeLinkHeaders) {
+    return headers
+  }
+
+  const { pages, manifest, pathPrefix, publicFolder } = pluginData
+  const perPageHeaders = preloadHeadersByPage({
+    pages,
+    manifest,
+    pathPrefix,
+    publicFolder,
+  })
+
+  return defaultMerge(headers, perPageHeaders)
+}
+
+const applySecurityHeaders = ({ mergeSecurityHeaders }) => headers => {
+  if (!mergeSecurityHeaders) {
+    return headers
+  }
+
+  return headersMerge(headers, SECURITY_HEADERS)
+}
+
+const applyCachingHeaders = (
+  pluginData,
+  { mergeCachingHeaders }
+) => headers => {
+  if (!mergeCachingHeaders) {
+    return headers
+  }
+
+  const chunks = Array.from(pluginData.pages.values()).map(
+    page => page.componentChunkName
+  )
+
+  chunks.push(`pages-manifest`, `app`)
+
+  const files = [].concat(...chunks.map(chunk => pluginData.manifest[chunk]))
```

Comment:
Why is this needed?

Reply:
We copied this over for the Netlify plugin. We'll review and iterate on this after release.